src/sage/schemes/elliptic_curves/addition_formulas_ring.py

```diff
@@ -0,0 +1,64 @@
+
+def add(E, P, Q):
+    r"""
+    Addition formulas for elliptic curves over general rings
+    with trivial Picard group.
```

Comment:
Add input and output blocks: https://doc.sagemath.org/html/en/developer/coding_basics.html#documentation-strings.

Can you include a brief explanation of why this function returns two values? I would expect an "add" function would only return one value, so this request for additional explanation is also partly to help me review the code.

Reply:
I'll add an explanation to the code momentarily. The point is that there are two separate addition formulas, which are equivalent whenever both work, but it is only guaranteed that some linear combination of them works for any given set of inputs. Hence why it yields the results for both possible formulas, and the caller will opportunistically take the first result that works, or combine both to obtain a valid projective point.